crates/polars-core/src/frame/group_by/mod.rs

```diff
@@ -40,15 +40,16 @@ impl DataFrame {
         let minimal_by_len = by.iter().map(|s| s.len()).min().expect("at least 1 key");
         let df_height = self.height();
 
-        // we only throw this error if self.width > 0
+        // Ensure all 'by' columns have the same height
+        // We only throw this error if self.width > 0
         // so that we can still call this on a dummy dataframe where we provide the keys
-        if (minimal_by_len != df_height) && (self.width() > 0) {
+        if (minimal_by_len != df_height || df_height == 0) && (self.width() > 0) {
             polars_ensure!(
-                minimal_by_len == 1,
+                minimal_by_len == 1 || df_height == 0,
                 ShapeMismatch: "series used as keys should have the same length as the DataFrame"
             );
             for by_key in by.iter_mut() {
-                if by_key.len() == minimal_by_len {
+                if by_key.len() != df_height {
```

Comment:
This isn't quite right. This will silently do the wrong thing if the minimal `by` column has length 1 but other columns have length != 1.